flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobClientHeartbeatHandler.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.rest.handler.job;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.api.common.time.Time;
+import org.apache.flink.runtime.rest.handler.AbstractRestHandler;
+import org.apache.flink.runtime.rest.handler.HandlerRequest;
+import org.apache.flink.runtime.rest.handler.RestHandlerException;
+import org.apache.flink.runtime.rest.messages.EmptyResponseBody;
+import org.apache.flink.runtime.rest.messages.JobClientHeartbeatParameters;
+import org.apache.flink.runtime.rest.messages.JobClientHeartbeatRequestBody;
+import org.apache.flink.runtime.rest.messages.JobIDPathParameter;
+import org.apache.flink.runtime.rest.messages.MessageHeaders;
+import org.apache.flink.runtime.webmonitor.RestfulGateway;
+import org.apache.flink.runtime.webmonitor.retriever.GatewayRetriever;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+/** Receive the heartbeat from the client. */
+public class JobClientHeartbeatHandler
+        extends AbstractRestHandler<
+                RestfulGateway,
+                JobClientHeartbeatRequestBody,
+                EmptyResponseBody,
+                JobClientHeartbeatParameters> {
+    public JobClientHeartbeatHandler(
+            GatewayRetriever<? extends RestfulGateway> leaderRetriever,
+            Time timeout,
+            Map<String, String> headers,
+            MessageHeaders<
+                            JobClientHeartbeatRequestBody,
+                            EmptyResponseBody,
+                            JobClientHeartbeatParameters>
+                    messageHeaders) {
+        super(leaderRetriever, timeout, headers, messageHeaders);
+    }
+
+    @Override
+    public CompletableFuture<EmptyResponseBody> handleRequest(
+            HandlerRequest<JobClientHeartbeatRequestBody> request, RestfulGateway gateway)
+            throws RestHandlerException {
+        final JobID jobId = request.getPathParameter(JobIDPathParameter.class);
+        gateway.reportJobClientHeartbeat(
+                jobId, request.getRequestBody().getExpiredTimestamp(), timeout);
```

Comment:
The client only needs to report its heartbeat and the dispatcher will deal with what to do when timeout. No need to return CompletableFuture since the client will not do any processing. What do you think?

Reply:
We probably don't need to return the CompletableFuture to the client side, but the rest handler should check the future returned from the gateway, and handle the exception if the future is completed exceptionally.